### PR TITLE
CLOUDSTACK-10239 default provider if needed

### PR DESCRIPTION
**Pull request by DaanHoogland**

### CODE REVIEW

**Comment by rhtyd** [comment]
@DaanHoogland does this PR solve the issue (blocker-ish) shared on users@ towards 4.11 RC1?

File: plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java [hunk 1]
```diff
@@ -54,11 +54,11 @@
     @Parameter(name = ApiConstants.TYPE, type = CommandType.STRING, required = true, description = "type of the ldap name. GROUP or OU")
     private String type;
 
-    @Parameter(name = ApiConstants.LDAP_DOMAIN, type = CommandType.STRING, required = true, description = "name of the group or OU in LDAP")
+    @Parameter(name = ApiConstants.LDAP_DOMAIN, type = CommandType.STRING, required = false, description = "name of the group or OU in LDAP")
```
Comment (rhtyd) [inline plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java]: Was this a regression that was added due to LDAP related enhancements, or could this cause issue if the field is made non-mandatory?

File: plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java [hunk 2]
```diff
@@ -54,11 +54,11 @@
...
     private String ldapDomain;
 
     @Deprecated
-    @Parameter(name = ApiConstants.NAME, type = CommandType.STRING, required = true, description = "name of the group or OU in LDAP")
+    @Parameter(name = ApiConstants.NAME, type = CommandType.STRING, required = false, description = "name of the group or OU in LDAP")
```
Comment (rhtyd) [inline plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java]: Same as previous comment ^^

File: plugins/user-authenticators/ldap/src/org/apache/cloudstack/ldap/LdapContextFactory.java [hunk 3]
```diff
@@ -80,9 +79,13 @@ private void enableSSL(final Hashtable<String, String> environment) {
 
     private Hashtable<String, String> getEnvironment(final String principal, final String password, final String providerUrl, final boolean isSystemContext, Long domainId) {
         final String factory = _ldapConfiguration.getFactory();
-        final String url = providerUrl == null ? _ldapConfiguration.getProviderUrl(domainId) : providerUrl;
+        String url = providerUrl == null ? _ldapConfiguration.getProviderUrl(domainId) : providerUrl;
+        if (StringUtils.isEmpty(url) && domainId != null) {
```
Comment (rhtyd) [inline plugins/user-authenticators/ldap/src/org/apache/cloudstack/ldap/LdapContextFactory.java]: Can we do a `Strings.isNullOrEmpty()` instead?

Comment (DaanHoogland) [reply plugins/user-authenticators/ldap/src/org/apache/cloudstack/ldap/LdapContextFactory.java]: we can but getProviderUrl(domainId) returns "" at least

File: plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java (hunk 1 above)
Comment (DaanHoogland) [reply plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java]: the validation of the field is done deeper in the stack. A check is done to see if either ldapdomain or name is filled.

File: plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java (hunk 2 above)
Comment (DaanHoogland) [reply plugins/user-authenticators/ldap/src/org/apache/cloudstack/api/command/LinkDomainToLdapCmd.java]: see above

**Comment by DaanHoogland** [comment]
@rhtyd it does,
@marcaurele I think this is in your toko, not?
I am not sure of my solution, it fixes backwards compatibility but I think the clean solution it to specify for which domain an ldap server is meant (as it can be specified for several domain anyway)

**Comment by DaanHoogland** [comment]
as it is a backwards compatibility issue, i am marking this as a blocker. happy to discuss though.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
Test LGTM, most of the the failures were caused by a fake public IP range added by test_public_ip_range.py that failed download urls for iso/volume/template.

**Pull request merged**